Sources/_CryptoExtras/RSA/RSA_security.swift

```diff
@@ -0,0 +1,318 @@
+                self = .rsaSignatureDigestPSSSHA256
+            case .pkcs1v1_5:
+                self = .rsaSignatureDigestPKCS1v15SHA256
+            }
+        case is SHA384.Digest.Type:
+            switch padding.backing {
+            case .pss:
+                self = .rsaSignatureDigestPSSSHA384
+            case .pkcs1v1_5:
+                self = .rsaSignatureDigestPKCS1v15SHA384
+            }
+        case is SHA512.Digest.Type:
+            switch padding.backing {
+            case .pss:
+                self = .rsaSignatureDigestPSSSHA512
+            case .pkcs1v1_5:
+                self = .rsaSignatureDigestPKCS1v15SHA512
+            }
+        default:
+            throw CryptoKitError.incorrectParameterSize
+        }
+    }
+}
+
+extension Data {
+    init<D: Digest>(_ digest: D) {
+        self = digest.withUnsafeBytes { Data($0) }
+    }
+
+    /// A partial PKCS8 DER prefix. This specifically is the version and private key algorithm identifier.
+    private static let partialPKCS8Prefix = Data(
+        [
+            0x02, 0x01, 0x00,  // Version, INTEGER 0
+            0x30, 0x0d,        // SEQUENCE, length 13
+            0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x01,  // rsaEncryption OID
+            0x05, 0x00         // NULL
+        ]
+    )
+
+    var pkcs8RSAKeyBytes: Data? {
+        // This is PKCS8. A bit awkward now. Rather than bring over the fully-fledged ASN.1 code from
+        // the main module and all its dependencies, we have a little hand-rolled verifier. To be a proper
+        // PKCS8 key, this should match:
+        //
+        // PrivateKeyInfo ::= SEQUENCE {
+        //   version                   Version,
+        //   privateKeyAlgorithm       PrivateKeyAlgorithmIdentifier,
+        //   privateKey                PrivateKey,
+        //   attributes           [0]  IMPLICIT Attributes OPTIONAL }
+        //
+        // Version ::= INTEGER
+        //
+        // PrivateKeyAlgorithmIdentifier ::= AlgorithmIdentifier
+        //
+        // PrivateKey ::= OCTET STRING
+        //
+        // Attributes ::= SET OF Attribute
+        //
+        // We know the version and algorithm identifier, so we can just strip the bytes we'd expect to see here. We do validate
+        // them though.
+        precondition(self.startIndex == 0)
+
+        guard self.count >= 4 + Data.partialPKCS8Prefix.count + 4 else {
+            return nil
+        }
+
+        // First byte will be the tag for sequence, 0x30.
+        guard self[0] == 0x30 else {
+            return nil
+        }
+
+        // The next few bytes will be a length. We'll expect it to be 3 bytes long, with the first byte telling us
+        // that it's 3 bytes long.
+        let lengthLength = Int(self[1])
+        guard lengthLength == 0x82 else {
+            return nil
+        }
+
+        let length = Int(self[2]) << 8 | Int(self[3])
+        guard length == self.count - 4 else {
+            return nil
+        }
```

Comment:
I don't follow this:

> We'll expect it to be 3 bytes long, with the first byte telling us that it's 3 bytes long.

If the first byte, `lengthLength` is `0x82`, how is that telling us it's 3 bytes long? And then why is `length` computed from only 2 bytes?

Reply:
`0x82` is an ASN.1-ism. It means: "the length of this field is definite, and the length field is the two octets after this one".